### ISSUE TITLE
[WebGPU] undefined symbol: wgpuGetProcAddress

### ISSUE THREAD

**Issue by emepetres** [emscripten-core/emscripten#11526]
Hi, I'm trying to call `wgpuGetProcAddress`, but the linker is telling me _undefined symbol_:

```
error: undefined symbol: wgpuGetProcAddress (referenced by top-level compiled C/C++ code)
warning: Link with `-s LLD_REPORT_UNDEFINED` to get more information on undefined symbols
warning: To disable errors for undefined symbols use `-s ERROR_ON_UNDEFINED_SYMBOLS=0`
warning: _wgpuGetProcAddress may need to be added to EXPORTED_FUNCTIONS if it arrives from a system library
Error: Aborting compilation due to previous errors
```

**Minimal sample that triggers the error**

`emcc -std=c++11 -s ENVIRONMENT=web -s WASM=1 -s USE_WEBGPU=1 test.cpp -o test.html`

```c++
//test.cpp
#include <emscripten/html5_webgpu.h>

int main() {
  auto device = emscripten_webgpu_get_device();
  auto func = wgpuGetProcAddress(device, "wgpuDeviceCreateRenderPipeline");
}
```

I'm using version 1.39.18. Is wgpuGetProcAddress not supported yet?

Thanks!

**Comment by emepetres** [comment]
Hi, could anyone take a look at this? Thanks :)

**Comment by kainino0x** [comment]
Sorry for the delay, I have a growing list of WebGPU work to do in Emscripten and I haven't gotten to any of it. I can probably take the time this week.

**Comment by kainino0x** [comment]
I'm not totally sure whether we will implement wgpuGetProcAddress in Emscripten - IIRC dynamic calls in Emscripten have a significant overhead (maybe it'll be disabled by default or something). What's your usecase that requires wgpuGetProcAddress instead of the statically loaded symbols?

**Comment by emepetres** [comment]
Thanks @kainino0x for looking at this.

We have developed a [WebGPU binding for .Net](https://github.com/WaveEngine/WebGPU.NET), and as in other bindings for [Vulkan](https://github.com/WaveEngine/Vulkan.NET) or [OpenGL](https://github.com/WaveEngine/OpenGL.NET) we are using the strategy of loading all API functions at binding initialization using getProcAddress. Each function pointer is then assigned to a delegate in C#.

This allows us to encapsulate better the binding from the specific implementation being used behind it.

As wgpuGetProcAddress would only be called on initialization, and only once per API function, I imagine that the overhead would be insignificant, am I right?

**Comment by sbc100** [comment]
> Thanks @kainino0x for looking at this.
> 
> We have developed a [WebGPU binding for .Net](https://github.com/WaveEngine/WebGPU.NET), and as in other bindings for [Vulkan](https://github.com/WaveEngine/Vulkan.NET) or [OpenGL](https://github.com/WaveEngine/OpenGL.NET) we are using the strategy of loading all API functions at binding initialization using getProcAddress. Each function pointer is then assigned to a delegate in C#.
> 
> This allows us to encapsulate better the binding from the specific implementation/architecture behind it.
> 
> As wgpuGetProcAddress would only be called on initialization, and only once per API function, I imagine that the overhead would be insignificant, am I right?

It seems like overhead would likely be something we would want to avoid.    I can see several overheads to doing things this way.   The biggest one for emscripten will be the lack ability to do DCE at link time across wasm and JS.   Any time you have a function like getProcAddress, it means the linker cannot eliminate any of the API surface, either on the native side or the JS side.    getProcAddress essentially keeps the whole API alive at link time.

Then are the lesser overheads, such as that fact that all calls end up going through and extra level indirection and the extra heap required api entry point (both for the string name of the function and the pointer its loaded into).  Admittedly these seems less likely be significant than the DCE problem.  We care a lot more about code size on the web than any other platform I know of these days.

@kripken might have more insight here into the history getProcAddress-type APIs in emscripten?

**Comment by kainino0x** [comment]
We should probably support it for completeness, but disabling by default could nudge people onto a better path and avoid the DCE problem for most users.

@juj may also have some input, since IIRC the GL code prefers static linking.

**Comment by kripken** [comment]
Yeah, these things are bad for DCE, and as indirect calls they are slower at runtime too. But, for compatibility, we've supported `SDL_GL_GetProcAddress` and others, using `emscripten_GetProcAddress`. So I think supporting this makes sense too, but we should not encourage people to use it.

**Comment by juj** [comment]
Agreed with @kripken on the direction. We can certainly add a `wgpuGetProcAddress` function, since it is easy to not use it whenever not needed - just don't call it and everything will DCE away without downsides.

Users of such `wgpuGetProcAddress` should be made well aware of the painful performance implications that it has: like Kai mentioned, referencing that function will cause the whole WebGPU API to be included in the output build code size, independent of whether all of the functions are called, bloating up the build; (and the size of the function `wgpuGetProcAddress` itself will not be small either) - but more importantly, calling WebGPU functions dynamically via the obtained function pointers will be noticeably slower on the CPU compared to calling WebGPU functions directly via static linkage, due to WebAssembly/JS environment security check/sandboxing reasons. Browsers will need to enforce that the dynamic function pointers are safe to call, creating a CPU-side micro-perf-hit, that has been quite noticeable single-digit % impact on WebGL API at least.

But users being aware of those downsides, no trouble with having a `wgpuGetProcAddress` for those that really need it.

**Comment by kainino0x** [comment]
> like Kai mentioned

(nit: Sam brought this up 🙂 )

**Comment by emepetres** [comment]
so I guess wgpuGetProcAddress won't be added after all...

**Comment by kainino0x** [comment]
We should do it, sorry I didn't triage the stalebot change.

I'm working on a bindings update in #14951 but won't get to this yet.

**Comment by kainino0x** [comment]
WebGPU bindings are moving into Dawn; migrating this bug to https://crbug.com/410607165
